src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,69 +3,107 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using System.Collections.Generic;
-using System.Diagnostics;
+using System.Buffers;
+using System.Buffers.Binary;
 using System.IO;
 using System.IO.Pipelines;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.Serialization;
-using Nerdbank.Streams;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
     internal static class RemoteHostAssetSerialization
     {
+        private static readonly ObjectPool<SerializableBytes.ReadWriteStream> s_streamPool = new(SerializableBytes.CreateWritableStream);
+
         public static async ValueTask WriteDataAsync(
-            Stream stream,
-            Dictionary<Checksum, object> assetMap,
-            ISerializerService serializer,
-            SolutionReplicationContext context,
-            Checksum solutionChecksum,
+            PipeWriter pipeWriter,
+            AssetPath assetPath,
             ReadOnlyMemory<Checksum> checksums,
+            SolutionAssetStorage.Scope scope,
+            ISerializerService serializer,
             CancellationToken cancellationToken)
         {
-            using var writer = new ObjectWriter(stream, leaveOpen: true, cancellationToken);
+            var pipeWriterStream = pipeWriter.AsStream();
 
-            // This information is not actually needed on the receiving end.  However, we still send it so that the
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+            var foundChecksumCount = 0;
 
-            // special case
-            if (checksums.Length == 0)
-                return;
+            await scope.AddAssetsAsync(
+                assetPath,
+                checksums,
+                WriteAssetToPipeAsync,
+                cancellationToken).ConfigureAwait(false);
```

Comment:
here's where we do the search for the assets corresponding to the passed in checksums.  As we find them, we call to WriteAssetToPipeAsync to immediately stream them to the server.